src/Storage/Device/S3.php

```diff
@@ -165,14 +169,124 @@ public function getPath(string $filename): string
      *
      * @param string $source
      * @param string $path
+     * @param int chunk
+     * @param int chunks
+     * @param array $metadata
      *
      * @throws \Exception
      *
+     * @return int
+     */
+    public function upload($source, $path, $chunk = 1, $chunks = 1, &$metadata = []): int
+    {
+        if($chunk == 1 && $chunks == 1) {
+            return $this->write($path, \file_get_contents($source), \mime_content_type($source));
+        }
+        $uploadId = $metadata['uploadId'] ?? null;
+        if(empty($uploadId)) {
+            $uploadId = $this->createMultipartUpload($path, $metadata['content_type']);
+            $metadata['uploadId'] = $uploadId;
+        }
+
+        $etag = $this->uploadPart($source, $path, $chunk, $uploadId);
+        $metadata['parts'] ??= [];
+        $metadata['parts'][] = ['partNumber' => $chunk, 'etag' => $etag];
+        $metadata['chunks'] ??= 0;
+        $metadata['chunks']++;
+        if($metadata['chunks'] == $chunks) {
+            $this->completeMultipartUpload($path, $uploadId, $metadata['parts']);
+        }
+        return $metadata['chunks'];
+    }
+
+    /**
+     * Start Multipart Upload
+     * 
+     * Initiate a multipart upload and return an upload ID.
+     * 
+     * @param string $path
+     * @param string $contentType
+     * 
+     * @return string
+     */
+    protected function createMultipartUpload(string $path, string $contentType): string
+    {
+        $uri = $path !== '' ? '/' . \str_replace(['%2F', '%3F'], ['/', '?'], \rawurlencode($path)) : '/';
+
+        $this->headers['date'] = gmdate('D, d M Y H:i:s T');
+        $this->headers['content-type'] = $contentType;
+        $this->amzHeaders['x-amz-acl'] = $this->acl;
+        $response = $this->call(self::METHOD_POST, $uri, '', ['uploads' => '']);
+        return $response->body['UploadId'];
+    }
+
+    /**
+     * Upload Part
+     * 
+     * @param string $source
+     * @param string $path
+     * @param string $uploadId
+     * @param int $chunk
```

Comment:
Nitpick: can the order in the docblock match the order of params?
```suggestion
     * @param int $chunk
     * @param string $uploadId
```